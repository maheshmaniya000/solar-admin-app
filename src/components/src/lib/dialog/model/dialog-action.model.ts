import { Observable } from 'rxjs';

export interface DialogAction<T = any> {
  dismissing: boolean;
  textTranslationKey: string;
  payload: T;
  disabled?: boolean;
  dataTest: string;
}

// https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-1.html#mapped-types-on-tuples-and-arrays
type TypesToDialogActions<Types extends any[]> = { [K in keyof Types]: DialogAction<Types[K]> };

export type DialogActions<PayloadTypes extends any[] = any[]> =
  | TypesToDialogActions<PayloadTypes>
  | Observable<TypesToDialogActions<PayloadTypes>>;
