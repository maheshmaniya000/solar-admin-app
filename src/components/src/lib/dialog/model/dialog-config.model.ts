import { MatDialogConfig } from '@angular/material/dialog';

export type DialogConfig<D = any> = Pick<MatDialogConfig<D>, 'data' | 'disableClose'>;
