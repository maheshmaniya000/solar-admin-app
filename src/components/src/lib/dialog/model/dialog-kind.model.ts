export enum DialogKind {
  medium = 'medium',
  large = 'large',
  featured = 'featured'
}
