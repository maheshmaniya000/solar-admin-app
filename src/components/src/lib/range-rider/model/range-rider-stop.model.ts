export interface RangeRiderStop {
  value: number;
  color: string;
}
