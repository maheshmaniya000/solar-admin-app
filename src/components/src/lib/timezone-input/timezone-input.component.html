<ng-container [formGroup]="form">
  <label for="timezone-select">Time zone</label>
  <mat-form-field doNotAppendFormError>
    <mat-select
      data-test="timezone-select"
      formControlName="timezone"
      id="timezone-select"
    >
      <mat-option
        *ngFor="let timezoneOption of timezoneOptions"
        [value]="timezoneOption.value"
      >
        {{ timezoneOption.label }}
      </mat-option>
    </mat-select>
  </mat-form-field>
  <mat-form-field>
    <input
      aria-label="utc-offset-input"
      data-test="utc-offset-input"
      formControlName="utcOffset"
      matInput
      max="14"
      min="-12"
      type="number"
    />
    <span matSuffix>h</span>
  </mat-form-field>
</ng-container>
