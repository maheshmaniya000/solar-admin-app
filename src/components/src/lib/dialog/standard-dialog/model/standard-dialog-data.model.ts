import { TemplateRef, Type } from '@angular/core';

import { DialogActions } from '../../model/dialog-action.model';

export interface StandardDialogData<PayloadTypes extends any[] = any[]> {
  content: string | TemplateRef<any> | Type<any>;
  titleTranslationKey?: string;
  closeIconButtonEnabled?: boolean;
  dividedContent?: boolean;
  actions?: DialogActions<PayloadTypes>;
  [key: string]: any;
}
