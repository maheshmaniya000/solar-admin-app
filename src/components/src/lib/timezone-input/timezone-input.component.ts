import {
  ChangeDetectionStrategy,
  Component,
  Input,
  OnChanges,
  OnDestroy,
  OnInit
} from '@angular/core';
import { AbstractControl, FormGroup, Validators } from '@angular/forms';
import { Subject } from 'rxjs';
import { startWith, takeUntil } from 'rxjs/operators';

import { SelectOption } from '../models/select-option.model';
import { Timezone } from '../models/timezone.model';
import { SgValidators } from '../validators/validators';

@Component({
  selector: 'sg-timezone-input',
  templateUrl: './timezone-input.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class TimezoneInputComponent implements OnDestroy, OnInit, OnChanges {
  @Input() form: FormGroup;
  @Input() timezone: Timezone;
  @Input() utcOffset: number;

  readonly timezoneOptions: SelectOption<Timezone>[] = [
    {
      label: 'Local legal time (rounded)',
      value: Timezone.site
    },
    {
      label: 'UTC offset',
      value: Timezone.utc
    }
  ];

  private readonly destroyed$ = new Subject<void>();

  static createControlConfigs(): { timezone: any; utcOffset: any } {
    return {
      timezone: [undefined, [Validators.required]],
      utcOffset: [
        undefined,
        [
          Validators.required,
          SgValidators.integer,
          Validators.min(-12),
          Validators.max(14)
        ]
      ]
    };
  }

  ngOnInit(): void {
    this.subscribeToTimezoneFormControlChanges();
  }

  ngOnChanges(): void {
    this.form.patchValue(
      { timezone: this.timezone, utcOffset: this.utcOffset },
      { emitEvent: false }
    );
    this.updateUtcOffsetFormControl(this.timezone);
  }

  ngOnDestroy(): void {
    this.destroyed$.next();
  }

  private getTimezoneFormControl(): AbstractControl {
    return this.form.get('timezone');
  }

  private getUtcOffsetFormControl(): AbstractControl {
    return this.form.get('utcOffset');
  }

  private subscribeToTimezoneFormControlChanges(): void {
    this.getTimezoneFormControl()
      .valueChanges.pipe(
        startWith(this.timezone as Timezone),
        takeUntil(this.destroyed$)
      )
      .subscribe(timezone => this.updateUtcOffsetFormControl(timezone));
  }

  private updateUtcOffsetFormControl(timezone: Timezone): void {
    const utcOffsetFormControl = this.getUtcOffsetFormControl();
    if (timezone === Timezone.site) {
      utcOffsetFormControl.setValue(this.utcOffset, { emitEvent: false });
      utcOffsetFormControl.disable({ emitEvent: false });
    } else {
      utcOffsetFormControl.enable({ emitEvent: false });
    }
  }
}
