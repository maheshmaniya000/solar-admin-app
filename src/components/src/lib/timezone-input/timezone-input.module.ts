import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { MatInputModule } from '@angular/material/input';
import { MatSelectModule } from '@angular/material/select';

import { FormErrorModule } from '../form-error/form-error.module';
import { FormFieldModule } from '../form-field/form-field.module';
import { TimezoneInputComponent } from './timezone-input.component';

@NgModule({
  declarations: [TimezoneInputComponent],
  imports: [
    CommonModule,
    ReactiveFormsModule,
    MatInputModule,
    MatSelectModule,
    FormFieldModule,
    FormErrorModule
  ],
  exports: [TimezoneInputComponent]
})
export class TimezoneInputModule {}
