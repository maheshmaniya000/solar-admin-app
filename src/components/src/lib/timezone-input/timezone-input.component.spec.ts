import { FormBuilder } from '@angular/forms';
import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';

import { Timezone } from '../models/timezone.model';
import { TimezoneInputComponent } from './timezone-input.component';

import spyOn = jest.spyOn;
import SpyInstance = jest.SpyInstance;

describe('TimezoneInputComponent', () => {
  describe('unit', () => {
    interface TimezoneForm {
      timezone: Timezone;
      utcOffset: number;
    }

    let component: TimezoneInputComponent;
    let updateValueAndValiditySpy: SpyInstance;

    beforeEach(() => {
      component = new TimezoneInputComponent();
      component.form = new FormBuilder().group({
        ...TimezoneInputComponent.createControlConfigs()
      });
      component.timezone = Timezone.site;
      component.utcOffset = 4;
      updateValueAndValiditySpy = spyOn(
        component.form,
        'updateValueAndValidity'
      );
      component.ngOnInit();
    });

    const patchForm = (formValue: TimezoneForm): void => {
      component.form.patchValue(formValue);
    };

    describe('ngOnInit', () => {
      it('should update form value and validity upon initialization', () => {
        expect(updateValueAndValiditySpy).toHaveBeenCalledTimes(2);
      });
    });

    describe('form', () => {
      it('should contain disabled utc offset form control when timezone = site', () => {
        patchForm({
          timezone: Timezone.site,
          utcOffset: 4
        });
        expect(component.form.get('utcOffset').disabled).toBe(true);
      });

      it('should contain enabled utc offset form control when timezone = utc', () => {
        patchForm({
          timezone: Timezone.utc,
          utcOffset: 12
        });
        expect(component.form.get('utcOffset')).toEqual(
          expect.objectContaining({
            disabled: false,
            value: 12
          })
        );
      });

      it('utc offset should be set to 4 (initial value) when switched to site timezone', () => {
        patchForm({
          timezone: Timezone.utc,
          utcOffset: 12
        });
        component.form.get('timezone').setValue(Timezone.site);
        expect(component.form.get('utcOffset')).toEqual(
          expect.objectContaining({
            disabled: true,
            value: 4
          })
        );
      });
    });

    describe('ngOnChanges', () => {
      it('should contain disabled utc offset form control when timezone = site', () => {
        component.timezone = Timezone.site;
        component.ngOnChanges();
        expect(component.form.get('utcOffset').disabled).toBe(true);
      });

      it('should contain enabled utc offset form control when timezone = utc', () => {
        component.timezone = Timezone.utc;
        component.ngOnChanges();
        expect(component.form.get('utcOffset')).toEqual(
          expect.objectContaining({
            disabled: false,
            value: 4
          })
        );
      });

      it('utc offset should be set to 4 (initial value) when switched to site timezone', () => {
        patchForm({
          timezone: Timezone.utc,
          utcOffset: 12
        });
        component.timezone = Timezone.site;
        component.ngOnChanges();
        expect(component.form.get('utcOffset')).toEqual(
          expect.objectContaining({
            disabled: true,
            value: 4
          })
        );
      });
    });
  });

  describe('component', () => {
    const createComponent = createComponentFactory({
      component: TimezoneInputComponent,
      shallow: true
    });
    let spectator: Spectator<TimezoneInputComponent>;

    beforeEach(() => {
      spectator = createComponent({
        props: {
          form: new FormBuilder().group({
            ...TimezoneInputComponent.createControlConfigs()
          }),
          timezone: Timezone.site,
          utcOffset: 4
        }
      });
    });

    it('should match snapshot', () => {
      expect(spectator.fixture).toMatchSnapshot();
    });
  });
});
