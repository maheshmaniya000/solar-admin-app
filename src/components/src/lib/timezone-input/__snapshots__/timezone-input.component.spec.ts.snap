// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimezoneInputComponent component should match snapshot 1`] = `
<sg-timezone-input
  __ngContext__={[Function LRootView]}
  destroyed$={[Function Subject]}
  form={[Function FormGroup]}
  timezone={[Function String]}
  timezoneOptions={[Function Array]}
  utcOffset={[Function Number]}
>
  <label
    for="timezone-select"
  >
    Time zone
  </label><mat-form-field
    donotappendformerror=""
  >
    <mat-select
      data-test="timezone-select"
      formcontrolname="timezone"
      id="timezone-select"
    >
      <mat-option>
         Local legal time (rounded) 
      </mat-option>
      <mat-option>
         UTC offset 
      </mat-option>
    </mat-select>
  </mat-form-field><mat-form-field>
    <input
      aria-label="utc-offset-input"
      data-test="utc-offset-input"
      formcontrolname="utcOffset"
      matinput=""
      max="14"
      min="-12"
      type="number"
    />
    <span
      matsuffix=""
    >
      h
    </span>
  </mat-form-field>
</sg-timezone-input>
`;
